src/coreclr/jit/layout.cpp

```diff
@@ -5,6 +5,64 @@
 #include "layout.h"
 #include "compiler.h"
 
+struct CustomLayoutKey
+{
+    unsigned    Size;
+    const BYTE* GCPtrTypes;
+
+    CustomLayoutKey(ClassLayout* layout)
+        : Size(layout->GetSize())
+        , GCPtrTypes(layout->m_gcPtrCount > 0 ? layout->GetGCPtrs() : nullptr)
+    {
+        assert(layout->IsCustomLayout());
+    }
+
+    CustomLayoutKey(const ClassLayoutBuilder& builder)
+        : Size(builder.m_size)
+        , GCPtrTypes(builder.m_gcPtrCount > 0 ? builder.m_gcPtrs : nullptr)
+    {
+    }
+
+    static bool Equals(const CustomLayoutKey& l, const CustomLayoutKey& r)
+    {
+        if (l.Size != r.Size)
+        {
+            return false;
+        }
+
+        if ((l.GCPtrTypes == nullptr) != (r.GCPtrTypes == nullptr))
+        {
+            return false;
+        }
+
+        if ((l.GCPtrTypes != nullptr) && (memcmp(l.GCPtrTypes, r.GCPtrTypes, l.Size / TARGET_POINTER_SIZE) != 0))
+        {
+            return false;
+        }
+
+        return true;
+    }
+
+    static unsigned GetHashCode(const CustomLayoutKey& key)
+    {
+        unsigned hash = key.Size;
+        if (key.GCPtrTypes != nullptr)
+        {
+            hash ^= 0xc4cfbb2a + (hash << 19) + (hash >> 13);
```

Comment:
Is there any science to these hex values?

Reply:
No, I just picked two random values here

Reply:
To be clear, the constant used below (0x9e3779b9) is a common one I've seen in hash functions. But here I wanted the hash to always include some information about `GCPtrTypes != nullptr` or `GCPtrTypes == nullptr`, so this is folded `<random constant> + 0x9e3779b9` for those two cases.